Update Plans blog post (#892)
pulumi/pulumi-hugo: themes/default/content/blog/announcing-public-preview-update-plans/index.md

---
title: "Announcing the public preview of Update Plans"
date: 2022-02-10T08:00:00-07:00 # TODO update this
meta_desc: Update Plans enable you to see and confirm the updates that will be made to your infrastructure and then apply those updates at a later time.

> meta_image: meta.png # TODO update this

Reviewer: Do we have somebody working on the pic?

Reviewer 2: Me, I've put one in now

authors:
    - fraser-waters
tags:
    - features
---

## Announcing the public preview of Update Plans

> Pulumi’s previews are an important part of any workflow where you want to see the changes that will be made to your infrastructure before actually making the changes (with `pulumi up`). However, today there is no guarantee that the `pulumi up` operation will do only what was previewed; if the program, or your infrastructure, changes between the preview and the update, the update might make additional changes to bring your infrastructure back in line with what’s defined in your program. We’ve [heard from many of you](https://github.com/pulumi/pulumi/issues/2318) that you need a strong guarantee about exactly which changes an update will make to your infrastructure, especially in critical and production environments.

Reviewer: > your infrastructure, changes between the preview and the update

Do update plans fix this? (drift of infra) I guess it might - in combination with refresh?

Today, I’m excited to announce the public preview of Update Plans, a new Pulumi feature that provides exactly this guarantee. Update Plans enable review and approval workflows for your infrastructure. Plans also help catch any unexpected changes that might happen between when you preview a change and when you apply that change. Update Plans work by saving the results of a `pulumi preview` to a _plan file_, which enables you to restrict subsequent `pulumi up` operations to only the actions saved in the plan file. This helps you ensure that what you saw in the `pulumi preview` is what will actually happen when you run `pulumi up`.

> 

Reviewer: Do we need a `<!--more-->` cut line?

{{< asciicast id="466462" >}}

When updates fail to validate against the plan pulumi will print what constraint failed. In the example above the resource `urn:pulumi:dev::aws-ts-webserver::aws:ec2/instance:Instance::web-server-www` changed the value of the property `userData`.

## New scenario enabled by Update Plans: pull request validation workflows

Software development teams commonly use version control-based workflows to review and monitor the code that is added to a codebase. For example, many teams protect their primary branch (e.g. `main`, `master`, `trunk`) from direct code pushes and instead require a pull or merge request where the changes can be reviewed and continuous integration (CI) and tests can be run.

With Update Plans, this same workflow is now possible with infrastructure managed by Pulumi. Assuming you already have CI set up for your pull/merge requests, you can run `pulumi preview --save-plan PLAN-FILENAME` and output the resulting plan file as a CI artifact. Then, update the CI workflow that runs when changes are merged to run `pulumi up --plan PLAN-FILENAME`.

> ## Try Update Plans

Reviewer: Shall we mention implicit Update Plans somewhere (what happens on `pulumi up` by default, with `PULUMI_EXPERIMENTAL` on)?

---
title: "Announcing the public preview of Update Plans"
date: 2022-02-10T08:00:00-07:00 # TODO update this
meta_desc: Update Plans enable you to see and confirm the updates that will be made to your infrastructure and then apply those updates at a later time.
meta_image: meta.png # TODO update this
authors:

>     - alex-mullans

Reviewer 3: ```suggestion
```

Reviewer 2: I wasn't sure about this. He has left, but he did write most of this and I kinda felt credit due where credit's due?

---
title: "Announcing the public preview of Update Plans"
date: 2022-02-09
meta_desc: Update Plans enable you to see and confirm the updates that will be made to your infrastructure and then apply those updates at a later time.
meta_image: update_plans_preview.png
authors:
    - fraser-waters
tags:
    - features
---

Pulumi’s previews are an important part of any workflow where you want to see the changes that will be made to your infrastructure before actually making the changes (with `pulumi up`). However, today there is no guarantee that the `pulumi up` operation will do only what was previewed; if the program, or your infrastructure, changes between the preview and the update, the update might make additional changes to bring your infrastructure back in line with what’s defined in your program. We’ve [heard from many of you](https://github.com/pulumi/pulumi/issues/2318) that you need a strong guarantee about exactly which changes an update will make to your infrastructure, especially in critical and production environments.

<!--more-->

> Today, I’m excited to announce the public preview of Update Plans, a new Pulumi feature which guarantees that operations shown in `pulumi preview` will actually run on `pulumi up`. Plans also help catch any unexpected changes that might happen between when you preview a change and when you apply that change. Update Plans work by saving the results of a `pulumi preview` to a _plan file_, which enables you to restrict subsequent `pulumi up` operations to only the actions saved in the plan file. This helps you ensure that what you saw in the `pulumi preview` is what will actually happen when you run `pulumi up`.

Reviewer 4: ```suggestion
Today, I’m excited to announce the public preview of Update Plans, a new Pulumi feature which guarantees that operations shown in `pulumi preview` will run on `pulumi up`. Update Plans also help catch any unexpected changes that might happen between when you preview a change and when you apply that change. Update Plans work by saving the results of a `pulumi preview` to a _plan file_, which enables you to restrict subsequent `pulumi up` operations to only the actions saved in the plan file. This helps you ensure that what you saw in the `pulumi preview` is what will actually happen when you run `pulumi up`.
```

> Here’s an example of Update Plans in action. In this example we're creating a new AWS EC2 instance and associated security group. We first plan that change with `pulumi preview --save-plan plan.json`, we then examain the plan file to see the resource changes the plan expects to make, and finally run the an update constrained to the plan with `pulumi up --plan plan.json`.

Reviewer 4: ```suggestion
Here’s an example of Update Plans in action. In this example we're creating a new AWS EC2 instance and associated security group. First, we plan the change with `pulumi preview --save-plan plan.json`, and then we examine the plan file to see which resources the plan expects to change. Finally, we run an update that's constrained to the plan with `pulumi up --plan plan.json`.
```

{{< asciicast id="466347" >}}

> To get started, you only need to add `--save-plan <file>` to the preview command and `--plan <file>` to the up command. In the video above, we demonstrated a pretty trivial example where our program did what our plan expected, so everything just ran as normal. If we clear the infrastructure we just created and start again with the same plan but change the program slightly, to have a different `userData` value, we can see that Update Plans returns an error and blocks the change.

Reviewer 4: ```suggestion
To get started, you only need to add `--save-plan <file>` to the preview command and `--plan <file>` to the up command. In the video above, we demonstrated a pretty trivial example where our program did what our plan expected, so everything ran just as expected. But what if we changed our program to have a different `userData` value and ran it against the same plan? If we clear the infrastructure we just created and use the same plan with the changed program, then Update Plans will return an error and block the change.
```

---
title: "Announcing the public preview of Update Plans"
date: 2022-02-08 # TODO update this
meta_desc: Update Plans enable you to see and confirm the updates that will be made to your infrastructure and then apply those updates at a later time.
meta_image: update_plans_preview.png
authors:
    - fraser-waters
tags:
    - features
---

> ## Announcing the public preview of Update Plans

Reviewer: You should drop this header to avoid
<img width="719" alt="image" src="https://user-images.githubusercontent.com/1454008/152971488-bd67c792-84db-4679-9289-e8187d9c4714.png">
